File: UnityPlugin/Assets/Plugins/iOS/Frontend/Notify/FCMNotifierPlugin.swift
// ======================================================================
// Project Name    : ios_foundation
//
// Copyright © 2017 U-CREATES. All rights reserved.
//
// This source code is the property of U-CREATES.
// If such findings are accepted at any time.
// We hope the tips and helpful in developing.
// ======================================================================
import Foundation
import UIKit
import UserNotifications
import Firebase
import FirebaseMessaging
open class FCMNotifierPlugin: BaseNotifierPlugin, UNUserNotificationCenterDelegate, FIRMessagingDelegate {
    fileprivate static var instance: FCMNotifierPlugin?
    fileprivate override init() {}
    open static func getInstance() -> FCMNotifierPlugin! {
        if (nil == FCMNotifierPlugin.instance) {
            FCMNotifierPlugin.instance = FCMNotifierPlugin()
        }
        return FCMNotifierPlugin.instance!
    }
    open override func register() -> Void {
        FIRApp.configure()
        let messaging: FIRMessaging = FIRMessaging.messaging()
        let notificationCenter: NotificationCenter = NotificationCenter.default
        messaging.remoteMessageDelegate = self
        notificationCenter.addObserver(self, selector: #selector(self.tokenRefreshNotification), name: .firInstanceIDTokenRefresh, object: nil)
        if #available(iOS 10.0, *) {
            func callback(granted: Bool, error: Error?) -> Void {
                if (nil != error) {
                    print(TagPlugin.UNITY_PLUGIN_IDENTIFIER + error.debugDescription)
                } else {
                    let firInstance: FIRInstanceID = FIRInstanceID.instanceID()
                    let fcmToken: String? = firInstance.token()
                    let app: UIApplication = UIApplication.shared
                    app.registerForRemoteNotifications()
                    if (nil != fcmToken) {
                        print(TagPlugin.UNITY_PLUGIN_IDENTIFIER + "fcmToken::" + fcmToken!)
                    }
                }
                return
            }
            let userNotificationCenter: UNUserNotificationCenter = UNUserNotificationCenter.current()
            userNotificationCenter.delegate = self
            userNotificationCenter.requestAuthorization(options: [.alert, .badge, .sound], completionHandler: callback)
        } else {
            let settings: UIUserNotificationSettings = UIUserNotificationSettings(types: [.alert, .badge, .sound], categories: nil)
            let app: UIApplication = UIApplication.shared
            app.registerUserNotificationSettings(settings)
        }
        return
    }
    open func tokenRefreshNotification(_ notification: Notification) -> Void {
        let firInstance: FIRInstanceID = FIRInstanceID.instanceID()
        let messaging: FIRMessaging = FIRMessaging.messaging()
        let fcmToken: String? = firInstance.token()
        if (nil == fcmToken) {
            print(TagPlugin.UNITY_PLUGIN_IDENTIFIER + "fcmToken is null")
            return
        }
        func callback(error: Error?) -> Void {
            if (nil != error) {
                print(TagPlugin.UNITY_PLUGIN_IDENTIFIER + error.debugDescription)
            } else {
                print(TagPlugin.UNITY_PLUGIN_IDENTIFIER + "success connect to fcm")
            }
            return
        }
        messaging.disconnect()
        messaging.connect(completion: callback)
        print(TagPlugin.UNITY_PLUGIN_IDENTIFIER + "fcmToken::" + fcmToken!)
        return
    }
    public func applicationReceivedRemoteMessage(_ remoteMessage: FIRMessagingRemoteMessage) -> Void {
        print(TagPlugin.UNITY_PLUGIN_IDENTIFIER + remoteMessage.appData.debugDescription)
        return
    }
    @available(iOS 10.0, *)
    public func userNotificationCenter(_ center: UNUserNotificationCenter, willPresent notification: UNNotification, withCompletionHandler completionHandler: @escaping (UNNotificationPresentationOptions) -> Void) {
        print(TagPlugin.UNITY_PLUGIN_IDENTIFIER + "execute fcm notification.")
        let userInfo: [AnyHashable: Any] = notification.request.content.userInfo
        self.noitfy(userInfo: userInfo)
        completionHandler([.alert, .badge, .sound])
        return
    }
    @available(iOS 10.0, *)
    public func userNotificationCenter(_ center: UNUserNotificationCenter, didReceive response: UNNotificationResponse, withCompletionHandler completionHandler: @escaping () -> Swift.Void) {
        print(TagPlugin.UNITY_PLUGIN_IDENTIFIER + "press fcm notification message in notify area.")
        let userInfo: [AnyHashable: Any] = response.notification.request.content.userInfo
        self.noitfy(userInfo: userInfo)
        completionHandler()
        return
    }
}

File: UnityAndroidPlugin/plugin/src/main/java/com/frontend/notify/FCMNotifierPlugin.java
// ======================================================================
// Project Name    : android_foundation
//
// Copyright © 2017 U-CREATES. All rights reserved.
//
// This source code is the property of U-CREATES.
// If such findings are accepted at any time.
// We hope the tips and helpful in developing.
// ======================================================================
package com.frontend.notify;
import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.util.Log;
import android.view.Window;
import android.view.WindowManager;
import com.core.identifier.TagPlugin;
import com.frontend.activity.ActivityPlugin;
import com.google.firebase.iid.FirebaseInstanceId;
public class FCMNotifierPlugin extends BaseNotifierPlugin {
    private static FCMNotifierPlugin instance = null;
    private FCMNotifierPlugin() {}
    public static FCMNotifierPlugin getInstance() {
        if (null == FCMNotifierPlugin.instance) {
            FCMNotifierPlugin.instance = new FCMNotifierPlugin();
        }
        return FCMNotifierPlugin.instance;
    }
    @Override
    public void register() {
        final Activity activity = ActivityPlugin.getInstance();
        Runnable runnable = new Runnable() {
            @Override
            public void run() {
                Window window = activity.getWindow();
                window.setFlags(WindowManager.LayoutParams.FLAG_FORCE_NOT_FULLSCREEN, WindowManager.LayoutParams.FLAG_FORCE_NOT_FULLSCREEN);
                FirebaseInstanceId fcmIId = FirebaseInstanceId.getInstance();
                String fcmToken = fcmIId.getToken();
                Log.i(TagPlugin.UNITY_PLUGIN_IDENTIFIER, "FCMToken::" + fcmToken);
                enable = true;
                return;
            }
        };
        activity.runOnUiThread(runnable);
        return;
    }
    @Override
    public void register(Context context, Intent intent) {
        String udt = intent.getStringExtra("google.c.a.udt");
        this.context = context;
        if (null != udt) {
            this.enable = true;
        }
        return;
    }
}
